File: planning-pattern/python/executor.py
from typing import Callable, Dict, Any, List

def execute_plan(steps: List[Dict[str, str]], on_progress: Callable[[int, str], None] | None = None) -> Dict[str, Any]:
    results: Dict[str, Any] = {}
    n = len(steps)
    for i, s in enumerate(steps):
        if on_progress:
            on_progress(round((i / max(n, 1)) * 100), s.get("id", ""))
        desc = s.get("description", "")
        if "Load numbers" in desc:
            results[s.get("id", "")] = [1, 2, 3, 4]
        elif "Compute average" in desc:
            arr = results.get("s1", [])
            results[s.get("id", "")] = (sum(arr) / len(arr)) if arr else None
        else:
            results[s.get("id", "")] = None
    if on_progress:
        on_progress(100, "done")
    return results

File: planning-pattern/typescript/src/executor.ts
export type ExecutionFailure = {
  status: "failed";
  error_type: "unsupported_step" | "missing_numbers";
  step_id: string;
  description: string;
};

export type ExecutionValue = number[] | number | ExecutionFailure;
export type ExecutionResults = Record<string, ExecutionValue>;

export async function executePlan(steps: { id: string; description: string }[], onProgress?: (pct: number, stage: string) => void): Promise<ExecutionResults> {
  const results: ExecutionResults = {};
  for (let i = 0; i < steps.length; i++) {
    const s = steps[i];
    onProgress?.(Math.round((i / steps.length) * 100), s.id);
    // trivial synthetic execution
    if (s.description.includes('Load numbers')) {
      const raw = s.description.match(/\[([^\]]+)\]/)?.[1] ?? '';
      results[s.id] = raw
        .split(',')
        .map(value => Number(value.trim()))
        .filter(Number.isFinite);
    }
    else if (s.description.includes('Compute average')) {
      const arr = Array.isArray(results['s1']) ? results['s1'] : [];
      results[s.id] = arr.length > 0
        ? arr.reduce((a:number,b:number)=>a+b,0)/arr.length
        : {
          status: "failed",
          error_type: "missing_numbers",
          step_id: s.id,
          description: s.description
        };
    } else results[s.id] = {
      status: "failed",
      error_type: "unsupported_step",
      step_id: s.id,
      description: s.description
    };
  }
  onProgress?.(100, 'done');
  return results;
}
